x-pack/plugins/fleet/server/routes/package_policy/handlers.ts

```diff
@@ -43,16 +43,74 @@ import { simplifiedPackagePolicytoNewPackagePolicy } from '../../../common/servi
 
 import type { SimplifiedPackagePolicy } from '../../../common/services/simplified_package_policy_helper';
 
-export const getPackagePoliciesHandler: RequestHandler<
+const getAllowedPackageNamesMessage = (allowedPackageNames: string[]): string => {
+  return `Allowed package.name's: ${allowedPackageNames.join(', ')}`;
+};
+
+/**
+ * Validates that Package Policy data only includes `package.name`'s that are in the list of
+ * `allowedPackageNames`. If an error is encountered, then a message is return, otherwise, undefined.
+ *
+ * @param data
+ * @param allowedPackageNames
+ */
+const validatePackagePolicyDataIsScopedToAllowedPackageNames = (
+  data: PackagePolicy[],
+  allowedPackageNames: string[]
+): string | undefined => {
+  if (!data.length) {
+    return;
+  }
+
+  if (!allowedPackageNames.length) {
+    return 'Authorization denied due to lack of integration package privileges';
+  }
+
+  // Because List type of APIs have an un-bounded `perPage` query param, we only validate the
+  // data up to the first package.name that we find is not authorized.
+  for (const packagePolicy of data) {
+    if (!packagePolicy.package) {
+      return `Authorization denied. ${getAllowedPackageNamesMessage(allowedPackageNames)}`;
+    }
+
+    if (!allowedPackageNames.includes(packagePolicy.package.name)) {
+      return `Authorization denied to [package.name=${
+        packagePolicy.package.name
+      }]. ${getAllowedPackageNamesMessage(allowedPackageNames)}`;
+    }
+  }
+};
+
+export const getPackagePoliciesHandler: FleetRequestHandler<
   undefined,
   TypeOf<typeof GetPackagePoliciesRequestSchema.query>
 > = async (context, request, response) => {
-  const soClient = (await context.core).savedObjects.client;
+  const fleetContext = await context.fleet;
+  const soClient = fleetContext.epm.internalSoClient;
+  const limitedToPackages = fleetContext.limitedToPackages;
+
   try {
     const { items, total, page, perPage } = await packagePolicyService.list(
       soClient,
       request.query
     );
+
+    // specific to package-level RBAC
+    if (limitedToPackages && limitedToPackages.length) {
```

Comment:
this block seems to be repeated a few times, can we extract to a function?

Reply:
Good idea. We'll be using similar logic in other API routes (elastic/security-team/issues/5620). Where do you suggest we should place this function so it can be imported where needed?  I think we should also have a generic `validatePolicyDataIsScopedToAllowedPackageNames` besides `validatePolicyDataIsScopedToAllowedPackageNames` in the same place, that takes in the agent policy, package policy or epm data to compute this.

something like:
```json5
const validatePolicyDataIsScopedToAllowedPackageNames = (
   data:  PackagePolicy[]| AgentPolicy[] | PackageInfo ,
   allowedPackageNames: string[]): string | undefined
): string | undefined => {
// validate if data is scoped to package names
}

const validateLimitedPackages(
   limitedPackages: string[] | undefined, 
   data: PackagePolicy[] | AgentPolicy[] | PackageInfo 
): IKibanaResponse<any> | undefined => {
    const validationResult = validatePolicyDataIsScopedToAllowedPackageNames(
        data,
        limitedToPackages
      );
      if (validationResult) {
        return response.forbidden({
          body: {
            message: validationResult,
          },
        });
      }
}
```

Reply:
you could place it under `fleet/server/routes` or the closest common ancestor

Reply:
I cleaned up some of the duplicated logic instead 68e366aa4a4805db4337ee0371de118741eb8e3e.